### PR TITLE
Add Multi-selection and Removal at the Recent List Screen

### PR DESCRIPTION
**Pull request by xie-dongping**
Add multi-selection, action bar and the according logics

Resolve issue #411 and #343

### CODE REVIEW

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/RecentListFragment.java [hunk 1]
```diff
@@ -303,24 +325,29 @@ public void onBindViewHolder(final ViewHolder holder, final int position) {
...
             holder.itemView.setOnLongClickListener(new View.OnLongClickListener() {
                 @Override
                 public boolean onLongClick(View v) {
-                    String dbPath = currentItem.dbPath;
-                    DialogFragment df = new OpenActionsFragment();
-                    Bundle b = new Bundle();
-                    b.putString(OpenActionsFragment.EXTRA_DBPATH, dbPath);
-                    df.setArguments(b);
-                    df.show(((FragmentActivity)context).getSupportFragmentManager(), "OpenActions");
+                    AppCompatActivity appCompatActivity = (AppCompatActivity) context;
```
Comment (xie-dongping) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/RecentListFragment.java]: I'm not sure if it should be the way to do it, as I'm not familiar with the data binding idiomatics in Android. I would certainly love to hear suggests about it.

Comment (helloworld1) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/RecentListFragment.java]: Your concern is right. Converting context to the activity is not the right approach. There are two ways to deal with this problem:
1. In holder's constructor, you need to set a callback Predicate<Void> startSupportActionModeProxy
When you create the Adapter, you will need to pass in the startSupportActionModeProxy from Activity thus passing down all the way to holder.
Then you can create the startSupportActionModeProxy in Activity with startSupportActionMode(actionModeCallback)

2. A simpler and less verbose way is using LocalBroadcastManager, when you want to startSupportActionMode, you basically send a broadcast intent. In the Activity, you register the broadcastReceiver in onStart and unregister in onStop and listen to the intent action you defined. When you receive the intent to startSupportActionMode, just call startSupportActionMode(actionModeCallback)

3. If you want to use EventBus, you will need to add the dependency: http://greenrobot.org/eventbus/, and this is by far the simplest, but it requires you to add a library.
It is very similar to approach 2.

**Comment by xie-dongping** [comment]
Thanks for the suggestion! I'll do it this way after my vacation.

helloworld1 <notifications@github.com>于2016年9月21日周三 22:23写道：

**Comment by helloworld1** [comment]
Thanks! Enjoy your vacation!

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/RecentListFragment.java (hunk 1 above)
Comment (xie-dongping) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/RecentListFragment.java]: I finally got some time to look into this and try to solve this problem, and I found it hard to solve it elegantly. 

I tried to go with route 2, as it introduces no extra dependencies. 

But as my `ActionModeCallback` requires also have access to the `recentListAdapter`, I would have to do it in this way (please correct me if I got it wrong):

1. Instantiate the `ActionModeCallback` in `RecentListAdapter`, and send signal in the `setOnClickListener`
2. Make the `actionModeCallback` a `Parcelable`, and send it with the `Intent`
3. In the `AnyMemo` `Activity`, all the `startSupportActionMode(actionModeCallback)` by using the `ActionModeCall`within the `Intent`.

Or I could do it in this way:

1. Send the signal for ActionMode in `setOnClickListener` of `RecentListAdapter
2. Create an `ActionModeCallBack` in `AnyMemo` activity, `AnyMemo` `Activity` invokes `startSupportActionMode(actionModeCallback)`
3. Send the signal when the selection should be removed or cleared in `ActionModeCallBack`.
4. Add another two two `BroadCastReceiver` in the `RecentListAdapter` to handle the `selectionRemoved()` and `clearSelection()` respectively.

I would slightly prefer the second way. What do you think?

Comment (helloworld1) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/RecentListFragment.java]: I agree the second approach is better. The easier approach is using green robot eventbus instead LocalBroadcastManager. I am ok adding it as a dependency. http://greenrobot.org/eventbus/

**Comment by xie-dongping** [comment]
A new version with LocalBroadcastManager is done. I'm not sure if green robot eventbus would make it any easier to implement, so I stick with the vanilla way for signal binding.

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/RecentListFragment.java [hunk 2]
```diff
@@ -88,6 +107,26 @@ public void onAttach(Context activity) {
         super.onAttach(activity);
         mActivity = (Activity) activity;
         setHasOptionsMenu(true);
+        LocalBroadcastManager.getInstance(mActivity).registerReceiver(mClearSelectionReceiver,
+                new IntentFilter("recent-list-clear-selection"));
```
Comment (helloworld1) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/RecentListFragment.java]: Could you make the intent action as a constant that is access both here and in the adapter?

You can put it in the Util you created.

public static final String RECENT_LIST_CLEAR_SELECTION = "recent-list-clear-selection";

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/RecentListFragment.java [hunk 3]
```diff
@@ -352,6 +401,31 @@ public synchronized void setItems(List<RecentItem> items) {
             this.items.addAll(items);
             this.notifyDataSetChanged();
         }
+
+        private void updateActionMode() {
+            Intent intent = new Intent("recent-list-update-action-mode");
+            intent.putExtra("selectedItemCount", getSelectedItemCount());
```
Comment (helloworld1) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/RecentListFragment.java]: Also put selectedItemCount as a constant

File: app/src/main/java/org/liberty/android/fantastischmemo/utils/RecentListActionModeUtil.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    private BroadcastReceiver mStopActionMode;
+
+    @Inject
+    public RecentListActionModeUtil(BaseActivity activityContext) {
+        this.mContext = activityContext;
+    }
+
+
+    public void register() {
+        mStartActionMode = new BroadcastReceiver() {
```
Comment (helloworld1) [inline app/src/main/java/org/liberty/android/fantastischmemo/utils/RecentListActionModeUtil.java]: I don't think you need to initialize the broadcast receiver in register() method. You make it as a final field and initialize it in constructor (or in the field)

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/RecentListFragment.java [hunk 5]
```diff
@@ -352,6 +401,31 @@ public synchronized void setItems(List<RecentItem> items) {
...
+            LocalBroadcastManager.getInstance(context).sendBroadcast(intent);
+        }
+
+        public void stopActionMode() {
```
Comment (helloworld1) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/RecentListFragment.java]: Will you be able to put methods that send broadcasts into your RecentListActionModeUtil?

File: app/src/main/res/menu/action_menu.xml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<menu xmlns:android="http://schemas.android.com/apk/res/android"
+	xmlns:app="http://schemas.android.com/apk/res-auto">
+	<item
+		android:id="@+id/action_menu_remove"
+		android:icon="@drawable/delete"
+		android:orderInCategory="1"
+		app:showAsAction="ifRoom"
+		android:title="Delete" />
```
Comment (helloworld1) [inline app/src/main/res/menu/action_menu.xml]: Please use the @string/delete_text for the title instead

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/helper/SelectableAdapter.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    /**
+     * Clear the selection status for all items
+     */
+    public void clearSelection() {
+        List<Integer> selection = getSelectedItems();
+        selectedItems.clear();
+        for (Integer i : selection) {
+            notifyItemChanged(i);
```
Comment (helloworld1) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/helper/SelectableAdapter.java]: I think you should use notifyItemRemoved(i) here.

File: app/src/main/res/layout/open_screen_recent_item.xml [hunk 8]
```diff
@@ -1,53 +1,63 @@
...
         android:layout_height="wrap_content"
-        android:gravity="center_vertical" >
-        <TextView
-            android:id="@+id/recent_item_filename"
-            android:layout_width="match_parent"
+        android:orientation="horizontal"
+        android:paddingTop="20dip"
+        android:paddingBottom="20dip"
+        android:paddingLeft="10dip"
+        android:paddingRight="1dip">
```
Comment (helloworld1) [inline app/src/main/res/layout/open_screen_recent_item.xml]: Please use dp as unit (The same as dip but it is officially called dp). 
Also could you standardlize these values and put it in dimens.xml?

Also 1dp for the right padding seems negligible. Is it necessary?

File: app/src/main/res/layout/open_screen_recent_item.xml [hunk 9]
```diff
@@ -1,53 +1,63 @@
...
-        android:layout_height="wrap_content"
-        android:text="@string/more_text"
-        style="@style/Widget.AppCompat.Button.Borderless.Colored"/>
-</LinearLayout>
+    <View
+        android:id="@+id/selected_overlay"
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:background="@color/selected_item"
+        android:visibility="invisible" />
```
Comment (helloworld1) [inline app/src/main/res/layout/open_screen_recent_item.xml]: Is it better to be visibility "gone" or "invisible"?

**Comment by xie-dongping** [comment]
Thanks for the feedback! I'll address them later this week!

**Comment by helloworld1** [comment]
Thank you Dongping. If you don't have time to do it. Please let me know and I will pull in and address those issues myself.

**Comment by xie-dongping** [comment]
Sorry for the late reply! I was sort of busy these days, but I'll certainly address the issues, and give you a version this week.

However, if you are rushing for a release, then certainly it would be OK if you wish to do it.

**Comment by helloworld1** [comment]
No worry. I am not in a rush to release new version. I will need to fix dropbox first. Just keep in touch :)

File: app/src/main/java/org/liberty/android/fantastischmemo/utils/RecentListActionModeUtil.java (hunk 4 above)
Comment (xie-dongping) [inline app/src/main/java/org/liberty/android/fantastischmemo/utils/RecentListActionModeUtil.java]: I'm not sure if I understand it right. I would still have to have a register() and deregister() method for registering and unregistering, right?

**Comment by xie-dongping** [comment]
@helloworld1 
Haowen, what do you think about the version now?

**Comment by helloworld1** [comment]
Thank you for making all the changes! Happy holidays!

**Comment by xie-dongping** [comment]
Happy holidays!

**Pull request merged**